Assets/SuspensionCollider.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Unity.MLAgents.Integrations.Match3;
+using UnityEngine;
+
+public class SuspensionCollider : MonoBehaviour
+{
+    private MeshCollider _meshCollider;
+    private int _lastFrameCollision;
+    private WheelSuspension _wheelSuspension;
+
+
+    public float contactDepth; //How much the suspension is compressed/extended
+    public float lastcontactDepth;
+    private Rigidbody _rb;
+    public float stiffnes;
+    const float RaycastOffset = 0.01f;
+    const float PenetrationTolerance = 0.001f;
+    public float damper;
+    private float maxAcceleration = 0.0f;
+
+
+    public SuspensionCollider()
+    {
+        _lastFrameCollision = 0;
+    }
+
+
+    private void Start()
+    {
+        _meshCollider = GetComponent<MeshCollider>();
+        _wheelSuspension = GetComponentInParent<WheelSuspension>();
+        _rb = GetComponentInParent<Rigidbody>();
+        stiffnes = _wheelSuspension.stiffness;
+        damper = _wheelSuspension.damper;
+    }
+
+    public void CalculateContactdepth(Collider other)
+    {
+        lastcontactDepth = contactDepth;
+        contactDepth = -_wheelSuspension.extensionDistance;
+        transform.localPosition = new Vector3(0, contactDepth, 0);
+        if (other == null)
+        {
+            return;
+        }
+
+        bool significantOverlap;
+        RayCastCollision();
+        do
+        {
+            significantOverlap = PenetrativeCollision(other);
+        } while (significantOverlap);
+
+        if (contactDepth > 0)
+        {
+            
+            float speed = (contactDepth - lastcontactDepth) / Time.fixedDeltaTime;
+            var springAcceleration = Math.Min(contactDepth * stiffnes + speed * damper,6f);
+            // var springAcceleration = contactDepth * stiffnes + speed * damper;
+            if (maxAcceleration < springAcceleration)
+            {
+                maxAcceleration = springAcceleration;
+            }
+            // Debug.Log($"Speed:{speed} --- ConatactDepth:{contactDepth} --- MaxAxx:{springAcceleration} timeFrame:{Time.frameCount} Wheel:{_wheelSuspension.name}");
+
+            var worldSpringAcceleration = transform.TransformVector(new Vector3(springAcceleration, 0.0f, 0.0f));
+            _rb.AddForceAtPosition(worldSpringAcceleration, _wheelSuspension.displacementCollider.transform.position,
+                ForceMode.Acceleration);
+        }
+    }
+
+    public bool PenetrativeCollision(Collider other)
+    {
+        var ownBodyTransform = _meshCollider.transform;
+        var collisionBodyTransform = other.transform;
+        var significantOverlap = Physics.ComputePenetration(_meshCollider, ownBodyTransform.position,
```

Comment:
Die Variable `significantOverlap` hat irgendwie keine passende Bedeutung zum Namen. Erst beschreibt sie nur, ob es einen Overlap gibt aus `Physics.ComputePenetration` und danach wird sie neu über `MoveWheel` gesetzt, wo die Bedeutung dann passt. Den `ComputePenetration` Aufruf könnte man direkt in der Bedingung machen und am Ende der Methode `false` returnen

Reply:
Das if Statement wirkt sehr überladen, wenn Compute Penetration da drin ist. Wir haben aber Move Wheel und ComuptePenetration in verschiedene Variablen getrennt